pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/impl/UserService.java

```diff
@@ -0,0 +1,511 @@
+        responseMap.put("token", token);
+
+        logger.info("getUserProfileResponse() .usedId field is set");
+        responseMap.put("userId", claims.get("sub").toString());
+
+        logger.info("getUserProfileResponse() .email field is set");
+        responseMap.put("email", claims.get("email").toString());
+
+        logger.info("getUserProfileResponse() acceptedTOS is set");
+
+        boolean acceptedTOS = tosService.hasUserAcceptedLatest(claims.get("sub").toString());
+
+        responseMap.put("acceptedTOS", "" + acceptedTOS);
+
+        logger.info("getUserProfileResponse() expirationDate is set");
+        Date expirationDate = new Date(Calendar.getInstance().getTimeInMillis() + this.tokenExpirationTime);
+        responseMap.put("expirationDate", ZonedDateTime.ofInstant(expirationDate.toInstant(), ZoneOffset.UTC).toString());
+
+        logger.info("getUserProfileResponse() finished");
+        return responseMap;
+    }
+
+    public User getUserById(String userId) {
+        Optional<User> user = this.userRepository.findById(UUID.fromString(userId));
+        if (user.isEmpty()) {
+            logger.error("getUserById() cannot find user by UUID: {}", userId);
+            throw new IllegalArgumentException("Cannot find user by input UUID: " + userId);
+        }
+
+        return user.get();
+    }
+
+    public List<User> getAllUsers() {
+        return this.userRepository.findAll();
+    }
+
+    public List<User> addUser(List<User> users) {
+        return this.userRepository.saveAll(users);
+    }
+
+    /**
+     * This check is to prevent non-super-admin user to create/remove a super admin role
+     * against a user(include themselves). Only super admin user could perform such actions.
+     *
+     * <p>
+     * if operations not related to super admin role updates, this will return true.
+     * </p>
+     * <p>
+     * The logic here is checking the state of the super admin role in the input and output users,
+     * if the state is changed, check if the user is a super admin to determine if the user could perform the action.
+     *
+     * @param currentUser  the user trying to perform the action
+     * @param inputUser    the user that is going to be updated
+     * @param originalUser there could be no original user when adding a new user
+     * @return true if the user could perform the action, false otherwise
+     */
+    private boolean allowUpdateSuperAdminRole(
+            @NotNull User currentUser,
+            @NotNull User inputUser,
+            User originalUser) {
+
+        // if current user is a super admin, this check will return true
+        for (Role role : currentUser.getRoles()) {
+            for (Privilege privilege : role.getPrivileges()) {
+                if (privilege.getName().equals(AuthNaming.AuthRoleNaming.SUPER_ADMIN)) {
+                    return true;
+                }
+            }
```

Comment:
A set is a hash map. Looping through the values in a set to find a match is as silly as `map.keys().stream().filter(thing::equals).anyMatch();`. Instead, just use `contains`

Reply:
Even though this is a set I believe we can use contains. We are matching on the name and not the object itself.

Reply:
We could update the equals method of the object, but I would need to verify we are using it for other `contains` calls or other operations that depend on the equals method.